### DISCUSSION TITLE
Roadmapping: plan for migrating network stack to Rust

### DISCUSSION THREAD

**Discussion by robgjansen** [shadow/shadow#2729] [Show and tell]
### TCP modules we might want:
  - TCP state machine
    - API:
      - pass in packet header and needed args, such as buffer space available
      - returns action to take, such as: drop, ack, reset, retransmit
  - congestion control (vegas, reno, cubic, bbr)
  - retransmission (fast retrans, retrans timeout)
  - acknowledgement (selective, quick, delayed acks)

### Plan for migrating network code:
1. ~fix qdisc bug~
2. ~enable rust sockets in the network interface~
3. ~finish moving majority of TCP syscall handling to rust~
4. ~figure out how to support rust status listeners in the TCP socket~
5. ~figure out how to deal with sendto/recvfrom syscall handlers for TCP~
6. ~rewrite UDP sockets in rust~
7. ~rewrite network interface and qdisc, and remove `CompatSocket` objects~ (#3480)
8. begin rust TCP implementation

### Design sketch for trait-based packet states:
    https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=02d0eae5a60980fac0424df83e1cb0ee

```rust
mod with_traits {
    struct PacketStateA { }
    struct PacketStateB { }
    
    trait PacketState {
        fn receive(self: Box<Self>) -> Box<dyn PacketState>;
    }
    impl PacketState for PacketStateA {
        fn receive(self: Box<Self>) -> Box<dyn PacketState> {
            Box::new(PacketStateB{})
        }
    }
    impl PacketState for PacketStateB {
        fn receive(self: Box<Self>) -> Box<dyn PacketState> {
            self
        }
    }
    pub fn receive(opt: &mut Option<Box<dyn PacketState>>) {
        let s = opt.take().unwrap();
        *opt = Some(s.receive());
    }
}
```

**Comment by stevenengler** [comment]
We've so far added a tcp library to shadow (`src/lib/tcp/`), and added a tcp socket (`src/main/host/descriptor/socket/inet/tcp.rs`) which bridges the tcp library with shadow's syscall/file interface. We have not yet implemented things like acknowledgements or congestion control, and are in the process of adding those now.
